src/vision/status_quo/alan_finds_database_drops_hard.md

```diff
@@ -0,0 +1,84 @@
+
+Alan has just written a function that extends YouBuy to connect to a database when his application receives a request. The request returns all the data for a single user to a customized page on his website. After launching the beta version of his application, Alan begins receiving complaints that his website sometimes displays a page with an error message and other times it seems to work as intended. When the page with an error message appears the message reads: "Can not open a connection to the database, because a connection is already open".
+
+```rust=
+use sqlx::Connection;
+
+#[async_std::main]
+async fn main() -> Result<(), sqlx::Error> {
+    // Create a connection
+
+    let conn = SqliteConnection::connect("sqlite::memory:").await?;
+
+    // Make a simple query to return the given parameter
+    let row: (i64,) = sqlx::query_as("SELECT $1")
+        .bind(150_i64)
+        .fetch_one(&conn).await?;
+
+    assert_eq!(row.0, 150);
+
+    Ok(())
+}
+```
+
+Alan sees the reports coming in from his beta users, but hasn't received any alarms from his application. He realizes he must start debugging. His initial testing produces the customized page for the current user as expected, but after a number of attempts he sees the page with the error message. His application shows an error that says that there is already an open database handle.
+
+
+## Searching for the Solution
+
+
+Alan tries to figure out what happened from the logs, but is unable to do so. Alan decides to attempt to recreate the scenario by doing extra load testing and trying to emulate the increased traffic. After some time, Alan gets lucky and is able to recreate the error, but he still doesn't know what could be the cause. Alan turns to the documentation for the `sqlx` crate to see if there are flags that might enable extra instrumentation but he can't find any (XXX check what crate offers).
+
+Alan, next, takes the following steps to determine the most likely source to explain the problem behavior in his application:
+
+1. Check the `sqlx` documentation for possible flags to explicitly close the connection but none are found.
+2. Examine the source code to figure out where the connection gets closed.
+3. Discover that the connection is closed through the destructor, Drop::drop only when it gets cleaned up aftering going out of scope.
+
+Alan realizes he must seek help from someone more experienced and knowledgable. He goes to his friend Barbara, and asks, "How can I get a log statement when the destructor runs?". Barbara tells him to wrap his database handle in another struct which has a destructor and to insert a log into the destructor for that struct we'll call `Foo`. Alan implements `Foo` and sees the new log when the destructor runs in the output. Now that the log works Alan runs program with full load in order to reproduce the problem. Alan analyzes the logs and notices that the new connection is being created and panicing before the destructor has run.
```

Comment:
it might be nice to have code examples, I imagine something like

```rust
struct MyConnection {
    connection: SqliteConnection
}

impl Drop for MyConnection {
    fn drop(&mut self) {
        log("dropping the connection");
    }
}
```

But I just realized, this won't actually work! this will print right away, but the *connection itself* doesn't get dropped when the dtor runs, that's kind of the whole point!

Reply:
we might have to weak this part of the story.